align/pnr/hpwl.py

```diff
@@ -129,21 +173,105 @@ def calculate_HPWL_from_placement_verilog_d( placement_verilog_d, concrete_name,
+                    aux( ctn)
+            order.append( cn)
+        else:
+            found_leaves.add(cn)
+            order.append( cn)
+    aux(concrete_name)
+    return order, found_modules, found_leaves
 
-        local_HPWL = Mx-mx + My-my if mx is not None else 0
+def calculate_HPWL_from_placement_verilog_d_bottom_up( placement_verilog_d, concrete_name):
 
-        logger_fn( f"from netlist HPWL: {'/'.join(hnet)}: {local_HPWL}")
+    modules = { module['concrete_name']: module for module in placement_verilog_d['modules']}
+    instances = { (module['concrete_name'],instance['instance_name']): instance for module in placement_verilog_d['modules'] for instance in module['instances']}
 
-        HPWL += local_HPWL
+    leaf_terminals = defaultdict(list)
 
+    for leaf in placement_verilog_d['leaves']:
+        ctn = leaf['concrete_name']
+        for terminal in leaf['terminals']:
+            leaf_terminals[(ctn,terminal['name'])].append( to_center(terminal['rect']))
+
+    order, found_modules, found_leaves = compute_topoorder( modules, concrete_name)
+    logger.debug( f'{found_modules} {found_leaves} {order}')
+
+    net_bboxes = defaultdict(SemiPerimeter)
+    net_local_hpwls = {}
+
+    for k, v in leaf_terminals.items():
+        for r in v:
+            net_bboxes[k].addRect( r)
+
+    for cn in order:
+        if cn in found_modules:
+            logger.debug( f'Working on {cn}')
+            module = modules[cn]
+
+            local_hpwl = 0
+
+            local_a = set()
+            for instance in module['instances']:
+                ctn = instance['concrete_template_name']
+
+                if ctn in net_local_hpwls:
+                    local_hpwl += net_local_hpwls[ctn]
+
+                tr = Transformation( **instance['transformation'])
+                for fa in instance['fa_map']:
+                    f = fa['formal']
+                    a = fa['actual']
+                    local_a.add(a)
+                    r = net_bboxes[(ctn,f)].toList()
+
+                    new_r = tr.hitRect( Rect( *r)).canonical().toList()
+                    net_bboxes[ (cn,a)].addRect( new_r)
+
+            for a in local_a.difference(set( module['parameters'])):
+                net_hpwl = net_bboxes[(cn,a)].dist()
+                logger.debug( f'Accounting for hidden net {a} {net_hpwl}')
+                local_hpwl += net_hpwl
+            
+            net_local_hpwls[cn] = local_hpwl
+
+
+    assert order[-1] == concrete_name
+
+    cn = concrete_name
+    module = modules[cn]
+
+    HPWL = net_local_hpwls[cn]
+    for a in module['parameters']:
+        net_hpwl = net_bboxes[(cn,a)].dist()
+        logger.debug( f'Accounting for top-level net {a} {net_hpwl}')
+        HPWL += net_hpwl
+        
     return HPWL
+
+def calculate_HPWL_from_placement_verilog_d( placement_verilog_d, concrete_name, nets_d):
+    hpwl_top_down = calculate_HPWL_from_placement_verilog_d_top_down( placement_verilog_d, concrete_name, nets_d)
```

Comment:
Currently calculating it both top-down and bottom-up and showing they are the same.